[Clang] Fix : More Detailed "No expected directives found"

Sh0g0-1758 opened the pull request:
Fixes: #58290

**PR SUMMARY**: "Updated the error message to use the proper prefix when no expected directives are found by changing the hard coded expected in the message to a dynamic value in two error messages"

Sh0g0-1758 committed:
Fix : more detailed no expected directive message

Sh0g0-1758 committed:
ran git clang formatter

Sh0g0-1758 committed:
[Github Automation] : Add support for optional colon after /cherry-pick and /branch

Sh0g0-1758 commented:
![Screenshot from 2024-01-17 02-39-09](https://github.com/llvm/llvm-project/assets/114918019/2e5538d1-4541-4222-adcb-ede0f95960fc)
Here you can see a working implementation of the asked feature.

Sh0g0-1758 commented:
I have pushed code in this PR that aims to solve this [issue](https://github.com/llvm/llvm-project/pull/66540) also.

Sh0g0-1758 commented:
![Screenshot from 2024-01-17 02-48-15](https://github.com/llvm/llvm-project/assets/114918019/c1dcb8a2-f19e-4b05-bf3c-a497aa8753b3)
Here is a working implementation of the second issue that I aim to solve.

asl commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -9,7 +9,9 @@
-//===----------------------------------------------------------------------===//
+//===----------------------------------------------------------------------===//]
+
+// look into this file as well
```
Remove

asl commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -1404,7 +1404,7 @@ void ASTContext::InitBuiltinTypes(const TargetInfo &Target,
   }
 }
-
+// maybe change here also.
```
Ditto

asl commented on clang/lib/Frontend/VerifyDiagnosticConsumer.cpp:
```diff
@@ -1098,7 +1098,10 @@ void VerifyDiagnosticConsumer::CheckDiagnostics() {
     // source file(s) processed.
     if (Status == HasNoDirectives) {
-      Diags.Report(diag::err_verify_no_directives).setForceEmit();
+      // change here
```
Ditto

asl commented on llvm/utils/git/github-automation.py:
```diff
@@ -635,16 +635,16 @@ def execute_command(self) -> bool:
         """
         This function reads lines from STDIN and executes the first command
         that it finds.  The 2 supported commands are:
-        /cherry-pick commit0 <commit1> <commit2> <...>
```
Do not include unrelated things into a PR

Sh0g0-1758 replied:
Hello, I was assigned two issues and I have solved both of them to the best of my knowledge. And I pushed the patch for the second issue after solving the first one. Since I can't create a new PR when my first PR has not been merged yet, So I thought of just pushing it as I wanted to work on more issues. The issues are : [this](https://github.com/llvm/llvm-project/issues/58290#issuecomment-1886650144) and [this](https://github.com/llvm/llvm-project/issues/64803)

Sh0g0-1758 replied:
Since @tbaederr was involved in both the issues, I hope both the patches can be considered in a single PR.

asl replied:
You certainly could submit multiple PRs for separate changes.

asl reviewed (changes requested):
Ensure that PR does not contain unrelated things. See https://llvm.org/docs/Contributing.html#how-to-submit-a-patch for more information

Sh0g0-1758 committed:
Removed Unnecesary Comments

Sh0g0-1758 commented:
I have removed the unnecessary comments. @asl please do let me know whether I should submit different PRs for the issues. But I think that @tbaederr can review them both in this PR only. Also @asl , do let me know any other required changes.

asl commented on clang/include/clang/Basic/Diagnostic.h:
```diff
@@ -9,7 +9,7 @@
 /// Defines the Diagnostic-related interfaces.
 //
-//===----------------------------------------------------------------------===//
+//===----------------------------------------------------------------------===//]
```
still unrelated change. I would suggest you to carefully check all the code in the PR

Sh0g0-1758 replied:
done

Sh0g0-1758 committed:
Removed Unrelated Changes

Sh0g0-1758 commented:
@asl, I have checked all the files I changed and I have removed the unrelated changes and the changes related to another issue. Really sorry for this, I hope you can overlook it since this is my very first PR in LLVM.

Sh0g0-1758 committed:
refactor : Change position of '' and revert tests in indexer.test

Sh0g0-1758 commented on clang-tools-extra/clangd/test/indexer.test:
```diff
@@ -5,5 +5,5 @@
 # strip those.
 # RUN: clangd-indexer %t.cpp -- -Xclang -verify --save-temps -- 2>&1 | FileCheck %s
-# CHECK-NOT: error: no expected directives found: consider use of 'expected-no-diagnostics'
+# CHECK-NOT: error: no expected directives found: consider use of {{.*}}-no-diagnostics
```
ah yes, overlooked that. Will revert it.

Sh0g0-1758 commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -179,7 +179,7 @@ def err_verify_invalid_no_diags : Error<
     "%select{'expected-no-diagnostics' directive|other expected directives}0">;
 def err_verify_no_directives : Error<
-    "no expected directives found: consider use of 'expected-no-diagnostics'">;
+    "no expected directives found: consider use of '%0'-no-diagnostics">;
```
changed it to this norm and modified the test cases correspondingly.

Sh0g0-1758 commented:
@tbaederr, please review the changes.

tbaederr commented on clang/test/ARCMT/verify.m:
```diff
@@ -11,7 +11,7 @@
 // expected-error@-1 {{}}
 
-//      CHECK: error: no expected directives found: consider use of 'expected-no-diagnostics'
+//      CHECK: error: no expected directives found: consider use of '{{.*}}-no-diagnostics'
```
Do all these tests really fail if you leave "expected-no-diagnostics" in the expected output? Or can you just change that to a concrete value instead of the regex?

Sh0g0-1758 replied:
yes since I changed the error output value in the tds file, these tests needed to be changed with the regex.

tbaederr replied:
And what is the output now? Both RUN lines only pass `-verify` so the prefix shouldn't change?

Sh0g0-1758 replied:
Oh, I think I overlooked this. I then would have to revert the test changes. Please give me a confirmation to do so and also suggest whether or not I should write separate test cases for -verify=foo,bar for the change that this PR intends to implement.

tbaederr replied:
This test only passes `-verify` though, what's the new output and why doesn't the old expected output work?

Sh0g0-1758 replied:
the intended behavior is that when only `-verify` is passed and there are no diagnostics in the file then `expected-no-diagnostics` should be the error and when `-verify=foo` is passed and no such diagnostics is present in the file, then `foo-no-diagnostics` should be the error.

Sh0g0-1758 replied:
@tbaederr, I understand what you are trying to point out and I hope I rectified that in my new commit. I will add a screenshot of the working implementation also.

Sh0g0-1758 committed:
Fix : Added Support for -verify= and -verify={multi diagnostics}, revert changes in only -verify= test cases in verify files

Sh0g0-1758 commented:
Please see the desired behavior in this screenshot : 

![Screenshot from 2024-01-20 00-23-12](https://github.com/llvm/llvm-project/assets/114918019/c8c7dcca-a1eb-4ee5-8fe7-53c7c2ac99ca)

With `-verify=` we get `expected-no-diagnostics` and with `-verify=foo` we get `foo-no-diagnostics`

Sh0g0-1758 committed:
Ran clang formatter

Sh0g0-1758 commented:
gentle ping @tbaederr .

tbaederr commented:
The policy is to ping _once a week_. It's the weekend.

Sh0g0-1758 commented:
Ah really sorry. Pardon my enthu (ó﹏ò｡)

tbaederr commented on clang/lib/Frontend/VerifyDiagnosticConsumer.cpp:
```diff
@@ -1098,7 +1098,16 @@ void VerifyDiagnosticConsumer::CheckDiagnostics() {
+      std::string directives;
+      for (auto &Prefix : Diags.getDiagnosticOptions().VerifyPrefixes) {
+        directives = directives + Prefix + ",";
+      }
```
In `PrintUnexpected` we're simply using `Diags.getDiagnosticOptions().VerifyPrefixed.begin`.  How is this case different?

Sh0g0-1758 replied:
yes suppose multiple diagnostics were not found like given in the third case here : 

> ![Screenshot from 2024-01-20 00-23-12](https://private-user-images.githubusercontent.com/114918019/298168762-c8c7dcca-a1eb-4ee5-8fe7-53c7c2ac99ca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5NDU2NTIsIm5iZiI6MTcwNTk0NTM1MiwicGF0aCI6Ii8xMTQ5MTgwMTkvMjk4MTY4NzYyLWM4YzdkY2NhLWExZWItNGVlNS04ZmU3LTUzYzdjMmFjOTljYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEyMlQxNzQyMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZTRjNmIwYmU4MDhhMGNhZjc3NzE2NTgwYmUzNjRiZjEzNDE5YzhmNTJmYmNhZDM3ZDBmZTg5NTUwNGViNzc2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.CEGvxZlT56vu6qZ73y9ZSxX4d2a2FhH5hYEW3Wp6sUM)

then in the error generated, I give all the diagnostics as ```consider use of 'foo,alpha,bar,gamma-no-diagnostics'```. But if that is not the intended behavior or if multiple errors are expected to be emitted in response to multiple diagnostics not found, then that can also be done. Please specify which of the two is required.

Sh0g0-1758 replied:
I think the one I implemented is a better option as it batched the same types of errors into one and gave the programmer a concise view of the error.

tbaederr replied:
Well but I can't add a "foo,alpha,bar,gamma-no-diagnostics"(literally) comment to the test, right? That doesn't work?

Sh0g0-1758 replied:
Yes, I see your point. In that case I will change the code to now emit multiple errors and not batch them into one. so if you pass ```-verify=foo,bar``` and there are no such directives in the code then two errors with ```foo-no-diagnostics``` and ```bar-no-diagnostics``` will be generated. Please give me a confirmation to do so.

Sh0g0-1758 replied:
Considering this as resolved.

Sh0g0-1758 committed:
Updated Diagnostic error string builder logic

tbaederr commented on clang/lib/Frontend/VerifyDiagnosticConsumer.cpp:
```diff
@@ -1098,7 +1098,13 @@ void VerifyDiagnosticConsumer::CheckDiagnostics() {
-      Diags.Report(diag::err_verify_no_directives).setForceEmit();
+      std::string directives;
+      if (Diags.getDiagnosticOptions().VerifyPrefixes.empty()) {
+        directives = "expected";
```
I don't think this can be empty. Have you checked that this can actually happen?

Sh0g0-1758 replied:
yes, when there are no diagnostics passed in the verify, ie. when we pass ```-verify=``` only. In this case ```expected-no-directives``` should be the error output when no directives are present. and when we pass ```-verify=foo``` then ```foo-no-diagnostics``` should be the error when no foo directives are present. If I don't include this check, then with only ```-verify=``` it results in a segmentation fault.

Sh0g0-1758 replied:
any thoughts on this?

tbaederr replied:
I guess that's fine then.

tbaederr commented on clang/lib/Frontend/VerifyDiagnosticConsumer.cpp:
```diff
@@ -1098,7 +1098,13 @@ void VerifyDiagnosticConsumer::CheckDiagnostics() {
     // source file(s) processed.
     if (Status == HasNoDirectives) {
-      Diags.Report(diag::err_verify_no_directives).setForceEmit();
+      std::string directives;
```
```suggestion
      std::string Directive;
```

Sh0g0-1758 replied:
done.

tbaederr commented:
Adding Aaron since we're probably getting close.

Sh0g0-1758 committed:
Updated directives to Directive

Sh0g0-1758 committed:
retrigger checks

Sh0g0-1758 commented:
@asl @tbaederr @AaronBallman, can you please review this PR and merge it if everything seems fine.

Sh0g0-1758 committed:
Reformat Directive to Err_Directive to get checks passed

Sh0g0-1758 committed:
Revert

Sh0g0-1758 committed:
Revert back to original change

Sh0g0-1758 committed:
Ran clang formatter

jrtc27 commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -179,7 +179,7 @@ def err_verify_invalid_no_diags : Error<
     "%select{'expected-no-diagnostics' directive|other expected directives}0">;
 def err_verify_no_directives : Error<
-    "no expected directives found: consider use of 'expected-no-diagnostics'">;
+    "no expected directives found: consider use of '%0-no-diagnostics'">;
```
```suggestion
    "no %0 directives found: consider use of '%0-no-diagnostics'">;
```
no?

jrtc27 replied:
I guess elsewhere we refer to them as expected diagnostics, it's only in single quotes that we intend it to match the source?

Sh0g0-1758 replied:
this is just some formatting issue. I was following the directions of the original issue, where @tbaederr mentioned changing only the latter expected part of the string.

AaronBallman replied:
> I guess elsewhere we refer to them as expected diagnostics, it's only in single quotes that we intend it to match the source?

Correct, there are expected (English sense of the word) directives, and there are 'expected' (syntax sense of the word) directives. We want single quotes around the syntax part.

Sh0g0-1758 replied:
Considering this as resolved.

jrtc27 commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -179,7 +179,7 @@ def err_verify_invalid_no_diags : Error<
     "%select{expected|'expected-no-diagnostics'}0 directive cannot follow "
     "%select{'expected-no-diagnostics' directive|other expected directives}0">;
```
This looks like it needs templating?

Sh0g0-1758 replied:
Can you please elaborate on this.

AaronBallman replied:
I think @jrtc27 is saying that this diagnostic should also change to use `%0` similar to what you've done for `err_verify_no_directives` because it has the same issue with saying `'expected-no-diagnostics'`.

Sh0g0-1758 replied:
I see, that can be done easily but I think according to LLVM PR standards, its best to limit one PR for one issue. Maybe you can start it as a different issue and then I will submit a PR for the same.

AaronBallman replied:
I think this is the same issue -- in both cases, it's a hard-coded `expected-no-diagnostics` that is incorrect in the presence of a different prefix being used. Because of that, I'd still recommend doing it as part of this PR instead of as a separate one.

Sh0g0-1758 replied:
Okay, sure thing, I will drop the changes within a few hours then.

Sh0g0-1758 replied:
@AaronBallman, I don't think it needs templating. After a careful look I believe that the current implementation is for ```-verify``` only. Also the error says ```err_verify_invalid_no_diags``` which implicitly means that we are passing no ```-verify``` diagnostics in the first place. My point being in this case, there are no prefixes being used.

Sh0g0-1758 replied:
@AaronBallman , please correct me if I am wrong.

Endilll replied:
@Sh0g0-1758 Please have patience and don't ping people more often than once a wekk. Clang is very constrained on reviewer time of experienced people like Aaron.

Sh0g0-1758 replied:
Ah really sorry about that. Since this was pending for a few days and I wanted to push more PRs for the other issues without making a separate branch and rebasing them, I got a little impatient. Will surely keep this thing in mind from now on.

Endilll replied:
I definitely sympathize with you wanting to push more than one PR at a time, but the way to do it is to follow GitHub flow, specifically a part of it that says you should work in branches, and create your PRs from then: https://docs.github.com/en/get-started/using-github/github-flow#create-a-branch. This way PRs that touch different code can go through as many rounds of review as needed, without blocking other PRs.

Nagging reviewers won't get you anywhere, or even worse, deter reviewers from your PRs.

Sh0g0-1758 replied:
Yes I understand. I made a beginners mistake of not making a branch for the changes in this PR. Will surely keep that in mind from now on. Also I hope you can overlook my enthu, was not trying to nag any reviewers.

AaronBallman replied:
No worries about the extra pings (we'd rather see people "too" enthusiastic than disengaged!), just something to keep in mind for the future. :-)

The reason I think this diagnostic also needs to be updated is because of: https://godbolt.org/z/sEf4o8YY9 and https://godbolt.org/z/8oMWffoxY where it says `'expected-no-diagnostics'` in the diagnostic message. I think those should also say `'foo-no-diagnostics'` similar to how you've fixed it for the `no expected directives found` warning.

Sh0g0-1758 replied:
Thanks a lot for looking over it. 

I have made the required changes and added more tests for the same. Please take a look and suggest changes.

Sh0g0-1758 commented:
gentle ping @reviewers

AaronBallman commented on clang/lib/Frontend/VerifyDiagnosticConsumer.cpp:
```diff
@@ -1098,7 +1098,14 @@ void VerifyDiagnosticConsumer::CheckDiagnostics() {
     // source file(s) processed.
     if (Status == HasNoDirectives) {
-      Diags.Report(diag::err_verify_no_directives).setForceEmit();
+      std::string Err_Directive;
+      if (Diags.getDiagnosticOptions().VerifyPrefixes.empty()) {
+        Err_Directive = "expected";
+      } else {
+        Err_Directive = *Diags.getDiagnosticOptions().VerifyPrefixes.begin();
+      }
+      Diags.Report(diag::err_verify_no_directives).setForceEmit()
+          << Err_Directive;
```
```suggestion
      std::string ErrDirective;
      if (Diags.getDiagnosticOptions().VerifyPrefixes.empty())
        ErrDirective = "expected";
      else
        ErrDirective = *Diags.getDiagnosticOptions().VerifyPrefixes.begin();
      Diags.Report(diag::err_verify_no_directives).setForceEmit()
          << ErrDirective;
```
We might want to split this logic out into a helper function if we're also updating `err_verify_invalid_no_diags` at the same time.

Sh0g0-1758 replied:
Please let me know if we are doing this also in this PR. I think its best to raise a separate issue for the same as LLVM PR standards imply.

AaronBallman reviewed (commented):
Thank you for the fix! Can you please add more description to what is being fixed and how to the PR summary (as that becomes the commit message when we land the changes)?

I don't think a release note is necessary because this isn't really a user-facing feature (it's a feature for developers of Clang), but the changes should come with some test coverage, probably in `clang/test/Frontend/verify.c` or a new test similar to it.

Sh0g0-1758 committed:
Merge branch 'llvm:main' into main

Sh0g0-1758 committed:
Removed {} in the logic builder

Sh0g0-1758 committed:
Added test for changes

Sh0g0-1758 commented:
Hey @AaronBallman, I have added a test for the same in ```clang/test/Frontend/verify.c``` and made the necessary changes as well. I have added a short **PR SUMMARY** to the top-level comment as you instructed. As for changes in other error messages, I think its best to merge this PR and raise a separate PR for the same.

Sh0g0-1758 committed:
Refactored Tests

Sh0g0-1758 commented:
gentle ping regarding this  :

> @AaronBallman, I don't think it needs templating. After a careful look I believe that the current implementation is for `-verify` only. Also the error says `err_verify_invalid_no_diags` which implicitly means that we are passing no `-verify` diagnostics in the first place. My point being in this case, there are no prefixes being used.

Sh0g0-1758 commented:
gentle ping. Please check the mergeability of this PR.

Endilll commented:
> gentle ping. Please check the mergeability of this PR.

Friendly reminder of one ping per week policy.

Sh0g0-1758 committed:
Templated the error message err_verify_invalid_no_diags

Sh0g0-1758 committed:
Ran clang formatter

AaronBallman commented on clang/lib/Frontend/VerifyDiagnosticConsumer.cpp:
```diff
@@ -396,6 +396,15 @@ class VerifyDiagnosticConsumer::MarkerTracker {
   }
 };
 
+std::string DetailedErrorString(DiagnosticsEngine &Diags) {
+  std::string ErrDirective;
+  if (Diags.getDiagnosticOptions().VerifyPrefixes.empty())
+    ErrDirective = "expected";
+  else
+    ErrDirective = *Diags.getDiagnosticOptions().VerifyPrefixes.begin();
+  return ErrDirective;
+}
```
```suggestion
static std::string DetailedErrorString(const DiagnosticsEngine &Diags) {
  if (Diags.getDiagnosticOptions().VerifyPrefixes.empty())
    return "expected";
  return *Diags.getDiagnosticOptions().VerifyPrefixes.begin();
}
```
Some minor changes here but are mostly NFC.

Sh0g0-1758 replied:
Done

AaronBallman reviewed (commented):
Thank you, aside from some small suggestions, this LGTM

Sh0g0-1758 committed:
Refactored DetailedErrorString Function

Sh0g0-1758 committed:
Retrigger Checks

AaronBallman reviewed (approved):
LGTM, thank you!

AaronBallman commented:
I believe the concerns from @asl and @tbaederr have been addressed and we're fine to land this, but I'll land it tomorrow to give them both a chance to weigh-in and to give precommit CI a chance to finish.
